Enable pushdown of complex filters on partition columns

mbasmanova opened the pull request:
Updated HivePageSource to evaluate complex filters on partition columns. This fixes TestHiveIntegrationSmokeTest.testPrunePartitionFailure .

oerling commented:
So far so good. 

 

This represents a bet that there will not be cases of filter expressions over constant and non-constant columns. If this gets the tests passed, please merge.

 

I expect the general case will be very rare and thus the  bet of  delaying addressing this may be situationally appropriate.

 

 

 

From: Maria Basmanova <notifications@github.com> 
Sent: Wednesday, March 20, 2019 10:31 AM
To: prestodb/presto <presto@noreply.github.com>
Cc: oerling <erling@xs4all.nl>; Review requested <review_requested@noreply.github.com>
Subject: Re: [prestodb/presto] Enable pushdown of complex filters on partition columns (#12501)

 

@mbasmanova <https://github.com/mbasmanova>  requested your review on: #12501 <https://github.com/prestodb/presto/pull/12501>  Enable pushdown of complex filters on partition columns.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub <https://github.com/prestodb/presto/pull/12501#event-2217633452> , or mute the thread <https://github.com/notifications/unsubscribe-auth/Ap73z3yrNDHsuYu9w16oNOUc6h1V2oiEks5vYnBAgaJpZM4b_3hu> .

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -230,22 +242,45 @@ public boolean isFinished()
-                for (int fieldId = 0; fieldId < columnMappings.size(); fieldId++) {
+                int maxOutputChannel = -1;
+                for (int i = 0; i < outputChannels.length; i++) {
+                    if (outputChannels[i] == -1) {
```
Seems unnecessary?  Doesn't change the behavior if you remove it.

mbasmanova replied:
@tdcmeehan That's right. Will remove.

elonazoulay commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
+        int[] outputRows = new int[1];
+        int numRows = filterFunction.filter(new Page(blocks), outputRows, errorSet);
+        if (!errorSet.isEmpty()) {
+            filterOnPrefilledValuesException = Optional.of(errorSet.getFirstError(1));
```
Since getFirstError can return null should this be `Optional.ofNullable(errorSet.getFirstError(1));`

mbasmanova replied:
@elonazoulay getFirstError can return null only if error set is empty, which is not the case here.

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
     @Override
     public boolean pushdownFilterAndProjection(PageSourceOptions options)
     {
+        ImmutableList.Builder<FilterFunction> remainingFilterFunctions = ImmutableList.builder();
```
I think it would be clear if this were factored out into several methods:

1) `pushdownFilterAndProjectionToPartitionColumns` (or similar name)
2) A method to extract out `PageSourceOptions` that will be sent to the delegate

And then we call method 1, get the result from method 2, and then send it to the delegate.

tdcmeehan replied:
So you would need to be able to get both the remaining filter functions, and also whether or not it failed.  I wonder if this could be simplified into empty list, so whenever remaining filter functions returns empty list (as it would if pushdown failed), then we trivially don't attempt to delegate.

mbasmanova replied:
@tdcmeehan Here is what this might look like:

```
    @Override
    public boolean pushdownFilterAndProjection(PageSourceOptions options)
    {
        List<FilterFunction> remainingFilterFunctions = pushdownFilterOnPrefilledColumns(options);

        if (filterOnPrefilledValuesFailed) {
            return true;
        }

        this.outputChannels = Arrays.copyOf(options.getOutputChannels(), options.getOutputChannels().length);

        filterAndProjectPushedDown = delegate.pushdownFilterAndProjection(createPageSourceOptionsForDelegate(options, remainingFilterFunctions));
        return filterAndProjectPushedDown;
    }

    private List<FilterFunction> pushdownFilterOnPrefilledColumns(PageSourceOptions options)
    {
        ImmutableList.Builder<FilterFunction> remainingFilterFunctions = ImmutableList.builder();
        for (FilterFunction filterFunction : options.getFilterFunctions()) {
            Set<ColumnMappingKind> inputKinds = getInputKinds(filterFunction);
            if (!inputKinds.contains(PREFILLED)) {
                remainingFilterFunctions.add(filterFunction);
                continue;
            }

            if (inputKinds.size() > 1) {
                // TODO Pass necessary pre-filled values to the delegate so it can evaluate these filters
                throw new UnsupportedOperationException("Filters on a mix of partition and non-partition columns are not supported");
            }

            if (!filterFunction.isDeterministic()) {
                throw new UnsupportedOperationException("Non-deterministic filters on partition columns are not supported");
            }

            if (!evaluateFilterFunctionOnPrefilledValues(filterFunction)) {
                filterOnPrefilledValuesFailed = true;
                close();
                break;
            }
        }
        return remainingFilterFunctions.build();
    }
```

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -230,22 +242,45 @@ public boolean isFinished()
                 return null;
             }
+
+            if (dataPage.getPositionCount() > 0 && filterOnPrefilledValuesException.isPresent()) {
```
Hmm so I wonder why we would have a prefilled value exception if we didn't have any page positions?

mbasmanova replied:
@tdcmeehan Tim, consider the following query 

`select * from t where a > 0 and cast(ds as integer) > 0`

Table t has 2 columns: a is a regular column, ds is a partition column. `cast(ds as integer)` expression throws an exception because ds is a '2019-01-02' string. If table t has rows with a > 0, then the query should fail. Otherwise, it should succeed.

In the latter case, early evaluation of `cast(ds as integer)` produces an error, however, that error needs to be surfaces only if `a > 0` returns some rows. The error should be suppressed if `a > 0` returns nothing.

tdcmeehan replied:
Makes sense, thanks for the explanation.

tdcmeehan commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
+                throw new UnsupportedOperationException("Filters on a mix of partition and non-partition columns are not supported");
+            }
+
+            if (!evaluateFilterFunction(filterFunction)) {
```
Just double checking, this means skip the entire partition, right?  Or is it an error case?

mbasmanova replied:
Yes, this means skip the entire file.

mbasmanova commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveIntegrationSmokeTest.java:
```diff
@@ -3629,9 +3629,35 @@ public void testPrunePartitionFailure()
+        assertUpdate("CREATE TABLE test_partition_predicates " +
+                "WITH (partitioned_by = ARRAY['p']) AS " +
+                "SELECT 123 x, 'abc' p", 1);
+
```
@yingsu00 I added this query and it worked correctly. There were some bugs with NOT IN that were fixed along with large IN.

mbasmanova commented on presto-hive/src/main/java/com/facebook/presto/hive/HivePageSource.java:
```diff
@@ -752,18 +789,79 @@ public IntArrayList computeEligibleRowIds(Page page)
+                .collect(toImmutableSet());
+    }
+
+    private boolean evaluateFilterFunction(FilterFunction filterFunction)
```
@yingsu00 How about evaluateFilterFunctionOnPrefilledValues?

mbasmanova commented:
@elonazoulay @tdcmeehan @yingsu00 Elon, Tim, Ying, thank you for the feedback. I believe I addressed all of the comments. Would you take another look?

mbasmanova commented:
Superseded by #12533
